services/datamanager/builtin/capture_test.go

```diff
@@ -235,13 +198,146 @@ func TestDataCaptureEnabled(t *testing.T) {
 	}
 }
 
-func waitForCaptureFiles(captureDir string) {
+func TestSwitchResource(t *testing.T) {
+	captureDir := t.TempDir()
+	mockClock := clk.NewMock()
+	// Make mockClock the package level clock used by the dmsvc so that we can simulate time's passage
+	clock = mockClock
+
+	// Set up robot config.
+	config, deps := setupConfig(t, enabledTabularCollectorConfigPath)
+	config.CaptureDisabled = false
+	config.ScheduledSyncDisabled = true
+	config.CaptureDir = captureDir
+
+	// Build and start data manager.
+	dmsvc, r := newTestDataManager(t)
+	defer func() {
+		test.That(t, dmsvc.Close(context.Background()), test.ShouldBeNil)
+	}()
+
+	resources := resourcesFromDeps(t, r, deps)
+	err := dmsvc.Reconfigure(context.Background(), resources, resource.Config{
+		ConvertedAttributes: config,
+	})
+	test.That(t, err, test.ShouldBeNil)
+	passTimeCtx1, cancelPassTime1 := context.WithCancel(context.Background())
+	donePassingTime1 := passTime(passTimeCtx1, mockClock, captureInterval)
+
+	waitForCaptureFilesToExceedNFiles(captureDir, 0)
+	testFilesContainSensorData(t, captureDir, nil)
+
+	cancelPassTime1()
+	<-donePassingTime1
+
+	// Change the resource named arm1 to show that the data manager recognizes that the collector has changed with no other config changes.
+	for resource := range resources {
+		if resource.Name == "arm1" {
+			newResource := inject.NewArm(resource.Name)
+			newResource.EndPositionFunc = func(ctx context.Context, extra map[string]interface{}) (spatialmath.Pose, error) {
+				// Return a different value from the initial arm1 resource.
+				return spatialmath.NewPoseFromPoint(r3.Vector{X: 888, Y: 888, Z: 888}), nil
+			}
+			resources[resource] = newResource
+		}
+	}
+
+	err = dmsvc.Reconfigure(context.Background(), resources, resource.Config{
+		ConvertedAttributes: config,
+	})
+	test.That(t, err, test.ShouldBeNil)
+
+	initialData, err := getSensorData(captureDir)
+	test.That(t, err, test.ShouldBeNil)
+
+	passTimeCtx2, cancelPassTime2 := context.WithCancel(context.Background())
+	donePassingTime2 := passTime(passTimeCtx2, mockClock, captureInterval)
+
+	// Test that sensor data is captured from the new collector.
+	waitForCaptureFilesToExceedNFiles(captureDir, len(getAllFileInfos(captureDir)))
+	testFilesContainSensorData(t, captureDir, func(t *testing.T, sd []*v1.SensorData) {
```

Comment:
Would you instead want to call getAllFilePaths, check that they have roughly the same number of readings (or at least > 0), and that one file has sensor readings with one value (444) and the other with another (1)?

Reply:
Oh cool - is it guaranteed that the data gathered by the two different resources would be split into different files? If so I can do this test change! Is there a preferred way to decode the file contents back to []SensorData?

Reply:
If Equals returns false, we [close the existing collector](https://github.com/viamrobotics/rdk/blob/51ceb9b2b7cb81a8f131a99901c48d03d482058d/services/datamanager/builtin/builtin.go#L223-L224) which includes [flushing the BufferedWriter](https://github.com/viamrobotics/rdk/blob/51ceb9b2b7cb81a8f131a99901c48d03d482058d/data/collector.go#L67), then create a new collector (though to the same subdirectory since the component/method/everything is the same). New files are also created when we hit MaxFileSize, which is 64KB.

I would reuse [this](https://github.com/viamrobotics/rdk/blob/51ceb9b2b7cb81a8f131a99901c48d03d482058d/services/datamanager/builtin/capture_test.go#L54-L56) for getting all file paths and their SensorData.